common/src/main/java/com/wynntils/features/chat/ChatMentionFeature.java

```diff
@@ -61,69 +61,44 @@ private Pattern buildPattern() {
 
-        // process and append siblings
-        for (Component c : comp.getSiblings()) {
-            curr.append(rewriteComponentTree(c));
+        // If the part is null, it means that we did not get mentioned,
+        // or the mention is in two different parts, we likely don't want to mark it.
+        if (partFinding == null) {
+            return;
         }
 
-        return curr;
-    }
-
-    private MutableComponent rewriteMentions(MutableComponent curr, String text, Style style) {
-        Matcher match = mentionPattern.matcher(text);
-
-        // if we match then rewrite the component if there are no matches the function will just return
-        if (match.find()) {
-            // get the start of the string before any mention and set curr to be it
-            curr = Component.literal(text.substring(0, match.start())).withStyle(style);
-
-            // do the name of the first mention
-            curr.append(Component.literal(mentionColor.get() + match.group(0)));
+        if (markMention.get()) {
+            Matcher matcher = styledText.getMatcher(mentionPattern, PartStyle.StyleType.NONE);
 
-            // Store the point at which this match ended
-            int lastEnd = match.end();
+            if (!matcher.find()) {
+                WynntilsMod.error(
+                        "Found a part matching the mention pattern, but couldn't find a match in the string.");
+                return;
+            }
 
-            // Handle any additional mentions within the message
-            while (match.find()) {
-                // get the bit in between of matches
-                curr.append(Component.literal(text.substring(lastEnd, match.start())))
-                        .withStyle(style);
+            StyledText newText = styledText.splitAt(matcher.start()).splitAt(matcher.end());
 
-                // get the name and recolor it
-                curr.append(Component.literal(mentionColor.get() + match.group(0)));
+            StyledTextPart partMatching = newText.getPartMatching(mentionPattern, PartStyle.StyleType.NONE);
 
-                // set the starting point for the next mentions before variable
-                lastEnd = match.end();
+            if (partMatching == null) {
+                WynntilsMod.error(
+                        "Found a part finding the mention pattern, but couldn't find a match in the string after splitting.");
+                return;
             }
 
-            // finally add any text after the mentions back onto the component tree
-            curr.append(Component.literal(text.substring(lastEnd))).withStyle(style);
+            newText = newText.replacePart(
+                    partMatching, partMatching.withStyle(style -> style.withColor(mentionColor.get())));
+
+            MutableComponent newComponent = newText.getComponent();
+            e.setMessage(newComponent);
         }
 
-        return curr;
```

Comment:
If you move this above the markMention.get() check, you can flatten out the rest of the method by short-cutting out a return if markMention is not true. Less symmetry, perhaps, but also fewer nesting levels. You decide.

Reply:
Some people might want the sound but no mark.